{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intelligent systems for bioinformatics- Group 1\n",
    "\n",
    "This work is developed in the ambit of curricular unit intelligent systems for bioinformatics of the Bioinformatic Master by:\n",
    "- [Beatriz Santos](https://github.com/beatrizsantos1607)\n",
    "- [Duarte Velho](https://github.com/duartebred)\n",
    "- [Ricardo Oliveira](https://github.com/ricardofoliveira61)\n",
    "- [Rita Nobrega](https://github.com/ritanobrega00)\n",
    "- [Rodrigo Esperança](https://github.com/esperancaa)\n",
    "\n",
    "This work consists in the analysis of a dataset through the utilization of machine learning algorithms, recurring to Python as the programming language.\n",
    "The entire analysis is present on a Jupyter Notebook, organized in sections (explained later on) containing succinct explanations of the procedures and decisions taken throughout the analysis.\n",
    "\n",
    "For this work we selected the [GDSC1](https://tdcommons.ai/multi_pred_tasks/drugres) dataset. This dataset contains the wet lab IC50 for 208 drugs in 1000 cancer cells lines and can be used to design models that can predict drug response since the same compound can have different levels of responses in different patients. With this we aim to design a model that given a pair of drug and cell line genomics profile can predict the drug response and find the best drug to treat certain patient. In this dataset the RMD normalized gene expression was used for cancer lines and the SMILES for drugs. Y is the log normalized IC50.\n",
    "\n",
    "## Notebook sections\n",
    "### 1. Preprocessing and data exploration\n",
    "- Review of all documentation available about the dataset\n",
    "- Load the dataset and realize an exploratory analysis \n",
    "- Prepare the dataset with the generation and selection of features and treatment of the missing values \n",
    "\n",
    "This stage corresponds to the 1st section of the Notebook where:\n",
    "- The dataset must the described according to the documentation\n",
    "- Summarize the characteristics of the data through an exploratory analysis\n",
    "- Description of the preprocessing steps justifying the choices\n",
    "- Include graphics that represent the main characteristics of the dataset\n",
    "\n",
    "### 2. Non-supervised learning\n",
    "- Utilization of the adequate visualization and dimensionality reduction technique\n",
    "-  Application of clustering methods\n",
    "\n",
    "This stage corresponds to the section 2 of the Notebook where:\n",
    "- The results must be analyzed and the procedures explain\n",
    "\n",
    "### 3. Machine Learning\n",
    "- Compare the behavior of different models/methods of machine learning through the calculation of the performance metrics\n",
    "- Present the best model for the dataset\n",
    "\n",
    "This stage correspond to the section 3 of the notebook and all the results must be reported and analyzed in a critical way\n",
    "\n",
    "### 4. Deep Leaning\n",
    "- Utilization of deep learning methods similarly to the stage 3\n",
    "\n",
    "This stage correspond to the section 4 of the notebook and must report the results and have a critical analysis.\n",
    "\n",
    "All the packages used during this work are listed on the cell below\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tdc.multi_pred import DrugRes\n",
    "from tdc.metadata import get_task2category\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Preprocessing and data exploration\n",
    " The first stage of this work consists of describing the dataset that is going to be used along the project. This stage is crucial to understand the data, its structure, its quality, and potential analysis. Let's start by reviewing all the available documentation.\n",
    "\n",
    " ### 1.1 Documentation review and dataset description\n",
    " Recent studies have shown that alterations in cancer genomes influence the clinical response to anticancer therapies. Nowadays, genomic changes are used as molecular biomarkers to identify patients most likely to benefit from a treatment, however many cancer drugs in development or already in use have not been linked to a specific genomic marker that could guide their clinical use to diminish the time needed to treat a patient. \n",
    "\n",
    " The discovery of cancer genome as a potential biomarker was only possible due to the advances in the recent years in high-throughput technologies, in particular, DNA sequencing technologies, that allow the sequencing on a scale that was previously unthinkable. To explore the increased knowledge of cancer genomics, preclinical studies that link the genomic complexity of cancer with functional readouts such as drug sensitivity are required. For these studies, cancer lines derived from many naturally occurring cancer types are essential to mimic the tissue type and genomic context of the cancer, and they also provide an easy system for experimental manipulation for molecular biology and drug discovery. For this reason, several studies have used cancer cell lines to link pharmacological data with genomic information and helped define therapeutic biomarkers as well as to demonstrate that pharmacogenomic profiling in cancer cell lines can be used as a biomarker discovery platform to guide the development of new cancer therapies.\n",
    "\n",
    "The Genomics of Drug sensitivity in Cancer database, or GDSC for short, was designed to facilitate the study and understanding of the molecular features that influence drug response in cancer cell lines. The database holds datasets of drug sensitivity in cancer cells and links these data to detailed genomic information to facilitate the discovery of molecular biomarkers for drug response. These efforts are expected to, in a near future, provide a complete description of the genomic changes that occur in many cancer types\n",
    "and profound insights into the origins, evolution and progression of cancer.\n",
    "\n",
    "In order to download and load the dataset, we will use the `DrugRes` class from the `tdcommons.multi_pred` package. This package will facilitate data access to perform an initial analysis of the dataset, including its structure, format and content. The code presented below was used to download/load the dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "100%|██████████| 140M/140M [00:25<00:00, 5.49MiB/s] \n",
      "Loading...\n",
      "Done!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug_ID</th>\n",
       "      <th>Drug</th>\n",
       "      <th>Cell Line_ID</th>\n",
       "      <th>Cell Line</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>[3.23827250519154, 2.98225419469807, 10.235490...</td>\n",
       "      <td>2.395685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>ES3</td>\n",
       "      <td>[8.690197905033282, 3.0914731119366, 9.9924871...</td>\n",
       "      <td>3.140923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>ES5</td>\n",
       "      <td>[8.233101127037282, 2.82468731112752, 10.01588...</td>\n",
       "      <td>3.968757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>ES7</td>\n",
       "      <td>[8.33346622426757, 3.9667571228514302, 9.79399...</td>\n",
       "      <td>2.692768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>[8.39134072442845, 2.9683601858810698, 10.2606...</td>\n",
       "      <td>2.478678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177305</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-1040</td>\n",
       "      <td>[8.65368534780164, 2.9238748715081, 10.1278774...</td>\n",
       "      <td>5.353963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177306</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-407</td>\n",
       "      <td>[8.57966425274312, 2.77877087774424, 9.7680113...</td>\n",
       "      <td>4.820567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177307</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-61</td>\n",
       "      <td>[8.077115751588071, 2.78132536810578, 10.03805...</td>\n",
       "      <td>5.785978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177308</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-81</td>\n",
       "      <td>[7.7976988637889, 2.6408995410198797, 9.463400...</td>\n",
       "      <td>5.393454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177309</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-C5</td>\n",
       "      <td>[7.606903479683809, 3.15731861619302, 10.15314...</td>\n",
       "      <td>5.099131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>177310 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Drug_ID                                               Drug  \\\n",
       "0       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "1       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "2       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "3       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "4       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "...           ...                                                ...   \n",
       "177305      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177306      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177307      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177308      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177309      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "\n",
       "       Cell Line_ID                                          Cell Line  \\\n",
       "0            MC-CAR  [3.23827250519154, 2.98225419469807, 10.235490...   \n",
       "1               ES3  [8.690197905033282, 3.0914731119366, 9.9924871...   \n",
       "2               ES5  [8.233101127037282, 2.82468731112752, 10.01588...   \n",
       "3               ES7  [8.33346622426757, 3.9667571228514302, 9.79399...   \n",
       "4             EW-11  [8.39134072442845, 2.9683601858810698, 10.2606...   \n",
       "...             ...                                                ...   \n",
       "177305     SNU-1040  [8.65368534780164, 2.9238748715081, 10.1278774...   \n",
       "177306      SNU-407  [8.57966425274312, 2.77877087774424, 9.7680113...   \n",
       "177307       SNU-61  [8.077115751588071, 2.78132536810578, 10.03805...   \n",
       "177308       SNU-81  [7.7976988637889, 2.6408995410198797, 9.463400...   \n",
       "177309       SNU-C5  [7.606903479683809, 3.15731861619302, 10.15314...   \n",
       "\n",
       "               Y  \n",
       "0       2.395685  \n",
       "1       3.140923  \n",
       "2       3.968757  \n",
       "3       2.692768  \n",
       "4       2.478678  \n",
       "...          ...  \n",
       "177305  5.353963  \n",
       "177306  4.820567  \n",
       "177307  5.785978  \n",
       "177308  5.393454  \n",
       "177309  5.099131  \n",
       "\n",
       "[177310 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# download/load of the dataset and conversion to pandas.dataframe\n",
    "data = DrugRes(name = 'GDSC1')\n",
    "dataframe = data.get_data()\n",
    "# download/load the gene symbols of the transciptomic profile\n",
    "gene_symbols = data.get_gene_symbols()\n",
    "\n",
    "# dataframe visualization\n",
    "dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Dataset Statistics ---\n",
      "208 unique drugs.\n",
      "958 unique cell lines.\n",
      "177310 drug-cell line pairs.\n",
      "--------------------------\n"
     ]
    }
   ],
   "source": [
    "data.print_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The dataset download contains the cell line drug sensitivity data. This data was generated from ongoing high-throughput screening performed by the Cancer Genome Project at the Wellcome Trust Sanger Institute and the Center for Molecular Therapeutics at Massachusetts General Hospital.\n",
    "\n",
    "This database is composed by 5 columns. The first column contains the name of the drug used on the screening, while the second column contains the 1D structure of the drug represented as a SMILES string. The compounds used during the screening are anticancer therapeutics encompassing both targeted agents and cytotoxic chemotherapeutics that are approved drugs already in use in the clinic or drugs undergoing clinical development and in clinical trials. The targeted drugs cover a wide range of targets and processes implicated in cancer biology including receptor tyrosine kinase signalling, cell cycle control, DNA damage response and the cytoskeleton.\n",
    "\n",
    "The third column contains the name of the cell line used on the screening. The total collection of cell lines available for screening includes around 1000 different cancer cell lines that have been carefully selected to represent the spectrum of common and rare types of adult and childhood cancers of epithelial, mesenchymal and hematopoietic origin. The fourth column contains the raw mean difference (RMD) normalized gene expression profile data of the cell line when subjected to a specific anticancer drug in a vectorized format.\n",
    "\n",
    "The last column corresponds to the cell line drug sensitivity values, which were measured using fluorescence-based cell viability assays after 72 hours of drug treatment, and dose-response curves were fitted to fluorescence signal intensities over nine drug concentrations. However, the values reported on the dataset are the normalized log values of the half maximal inhibitory concentration (IC50).\n",
    "\n",
    "Is important to note that the current release of GDSC (3rd December of 2024) includes drug sensitivity data for 208 anticancer drugs and about 180000 cell line-drug interactions. This is the largest public resource available on drug sensitivity in cancer cells with ongoing screening with releases occurring every 4 months. Each new release these results are updated and new data for existing drug as well as data for new drugs screened is added.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading of extra information about the cell lines and drugs\n",
    "cell_lines = pd.read_csv(\"data/Cell_listTue Dec 3 15_39_17 2024.csv\")\n",
    "drugs = pd.read_csv(\"data/Drug_listTue Dec 3 15_39_27 2024.csv\")\n",
    "\n",
    "# filtering the dataset to only include cell lines and drugs that are in the current release of GDSC1\n",
    "## filtering cell lines\n",
    "cell_line_idx = []\n",
    "for cell_name in np.unique(dataframe[\"Cell Line_ID\"]):\n",
    "    idx = np.where(cell_lines[\"Cell line Name\"]==cell_name)[0]\n",
    "    if len(idx) != 0:\n",
    "        cell_line_idx.append(idx[0])\n",
    "\n",
    "cell_line_idx = np.array(cell_line_idx)\n",
    "cell_line = cell_lines.iloc[cell_line_idx]\n",
    "\n",
    "## filtering drugs\n",
    "drug_idx = []\n",
    "for drug_name in np.unique(dataframe[\"Drug_ID\"]):\n",
    "    idx = np.where(drugs[\" Name\"]==drug_name)[0]\n",
    "    if len(idx) != 0:\n",
    "        drug_idx.append(idx[0])\n",
    "\n",
    "drug_idx = np.array(drug_idx)\n",
    "drugs = drugs.iloc[drug_idx]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell line Name</th>\n",
       "      <th>Model ID</th>\n",
       "      <th>COSMIC ID</th>\n",
       "      <th>TCGA Classfication</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Tissue sub-type</th>\n",
       "      <th>Datasets</th>\n",
       "      <th>number of drugs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>SIDM00499</td>\n",
       "      <td>924100</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>urogenital_system</td>\n",
       "      <td>prostate</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>SIDM00980</td>\n",
       "      <td>910924</td>\n",
       "      <td>STAD</td>\n",
       "      <td>digestive_system</td>\n",
       "      <td>stomach</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>SIDM00982</td>\n",
       "      <td>687561</td>\n",
       "      <td>GBM</td>\n",
       "      <td>nervous_system</td>\n",
       "      <td>glioma</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>451Lu</td>\n",
       "      <td>SIDM01240</td>\n",
       "      <td>1287706</td>\n",
       "      <td>SKCM</td>\n",
       "      <td>skin</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5637</td>\n",
       "      <td>SIDM00807</td>\n",
       "      <td>687452</td>\n",
       "      <td>BLCA</td>\n",
       "      <td>urogenital_system</td>\n",
       "      <td>bladder</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1929</th>\n",
       "      <td>YT</td>\n",
       "      <td>SIDM00410</td>\n",
       "      <td>946358</td>\n",
       "      <td>ALL</td>\n",
       "      <td>blood</td>\n",
       "      <td>lymphoid_neoplasm_other</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1931</th>\n",
       "      <td>ZR-75-30</td>\n",
       "      <td>SIDM00971</td>\n",
       "      <td>909907</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>breast</td>\n",
       "      <td>breast</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1933</th>\n",
       "      <td>huH-1</td>\n",
       "      <td>SIDM00586</td>\n",
       "      <td>1298146</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>digestive_system</td>\n",
       "      <td>liver</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>no-10</td>\n",
       "      <td>SIDM00574</td>\n",
       "      <td>908452</td>\n",
       "      <td>LGG</td>\n",
       "      <td>nervous_system</td>\n",
       "      <td>glioma</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>no-11</td>\n",
       "      <td>SIDM00573</td>\n",
       "      <td>908450</td>\n",
       "      <td>LGG</td>\n",
       "      <td>nervous_system</td>\n",
       "      <td>glioma</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>390</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>941 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Cell line Name   Model ID   COSMIC ID  TCGA Classfication  \\\n",
       "0             22RV1  SIDM00499      924100                PRAD   \n",
       "2          23132-87  SIDM00980      910924                STAD   \n",
       "4          42-MG-BA  SIDM00982      687561                 GBM   \n",
       "6             451Lu  SIDM01240     1287706                SKCM   \n",
       "8              5637  SIDM00807      687452                BLCA   \n",
       "...             ...        ...         ...                 ...   \n",
       "1929             YT  SIDM00410      946358                 ALL   \n",
       "1931       ZR-75-30  SIDM00971      909907                BRCA   \n",
       "1933          huH-1  SIDM00586     1298146                LIHC   \n",
       "1935          no-10  SIDM00574      908452                 LGG   \n",
       "1937          no-11  SIDM00573      908450                 LGG   \n",
       "\n",
       "                 Tissue          Tissue sub-type  Datasets   number of drugs  \n",
       "0     urogenital_system                 prostate     GDSC1               353  \n",
       "2      digestive_system                  stomach     GDSC1               344  \n",
       "4        nervous_system                   glioma     GDSC1               345  \n",
       "6                  skin                 melanoma     GDSC1               352  \n",
       "8     urogenital_system                  bladder     GDSC1               347  \n",
       "...                 ...                      ...       ...               ...  \n",
       "1929              blood  lymphoid_neoplasm_other     GDSC1               387  \n",
       "1931             breast                   breast     GDSC1               341  \n",
       "1933   digestive_system                    liver     GDSC1               352  \n",
       "1935     nervous_system                   glioma     GDSC1               393  \n",
       "1937     nervous_system                   glioma     GDSC1               390  \n",
       "\n",
       "[941 rows x 8 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug Id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Targets</th>\n",
       "      <th>Target pathway</th>\n",
       "      <th>PubCHEM</th>\n",
       "      <th>Datasets</th>\n",
       "      <th>number of cell lines</th>\n",
       "      <th>Screening site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1242</td>\n",
       "      <td>(5Z)-7-Oxozeaenol</td>\n",
       "      <td>5Z-7-Oxozeaenol, LL-Z1640-2</td>\n",
       "      <td>TAK1</td>\n",
       "      <td>Other, kinases</td>\n",
       "      <td>9863776</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>899</td>\n",
       "      <td>SANGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>179</td>\n",
       "      <td>5-Fluorouracil</td>\n",
       "      <td>5-FU</td>\n",
       "      <td>Antimetabolite (DNA &amp; RNA)</td>\n",
       "      <td>Other</td>\n",
       "      <td>3385</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>907</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>86</td>\n",
       "      <td>A-443654</td>\n",
       "      <td>KIN001-139</td>\n",
       "      <td>AKT1, AKT2, AKT3</td>\n",
       "      <td>PI3K/MTOR signaling</td>\n",
       "      <td>10172943</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>397</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>55</td>\n",
       "      <td>A-770041</td>\n",
       "      <td>KIN001-111</td>\n",
       "      <td>LCK, FYN</td>\n",
       "      <td>Other, kinases</td>\n",
       "      <td>9549184</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>397</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1001</td>\n",
       "      <td>AICA Ribonucleotide</td>\n",
       "      <td>AICAR, N1-(b-D-Ribofuranosyl)-5-aminoimidazole...</td>\n",
       "      <td>AMPK agonist</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>65110</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>925</td>\n",
       "      <td>SANGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>310</td>\n",
       "      <td>YM201636</td>\n",
       "      <td>YM-201636, YM 201636</td>\n",
       "      <td>PIKFYVE</td>\n",
       "      <td>PI3K/MTOR signaling</td>\n",
       "      <td>9956222</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>912</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>45</td>\n",
       "      <td>Z-LLNle-CHO</td>\n",
       "      <td>Z-L-Norleucine-CHO, Gamma-Secretase Inhibitor 1</td>\n",
       "      <td>gamma-secretase</td>\n",
       "      <td>Other</td>\n",
       "      <td>16760646</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>397</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>1050</td>\n",
       "      <td>ZM447439</td>\n",
       "      <td>ZM-447439, ZM 447439</td>\n",
       "      <td>AURKA, AURKB</td>\n",
       "      <td>Mitosis</td>\n",
       "      <td>9914412</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>933</td>\n",
       "      <td>SANGER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>223</td>\n",
       "      <td>ZSTK474</td>\n",
       "      <td>KIN001-167, ZSTK-474, ZSTK 474</td>\n",
       "      <td>PI3K (class 1)</td>\n",
       "      <td>PI3K/MTOR signaling</td>\n",
       "      <td>11647372</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>911</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>266</td>\n",
       "      <td>Zibotentan</td>\n",
       "      <td>ZD4054, ZD-4054</td>\n",
       "      <td>Endothelin-1 receptor (EDNRA)</td>\n",
       "      <td>Other</td>\n",
       "      <td>9910224</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>912</td>\n",
       "      <td>MGH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Drug Id                 Name  \\\n",
       "0       1242    (5Z)-7-Oxozeaenol   \n",
       "4        179       5-Fluorouracil   \n",
       "20        86             A-443654   \n",
       "21        55             A-770041   \n",
       "29      1001  AICA Ribonucleotide   \n",
       "..       ...                  ...   \n",
       "683      310             YM201636   \n",
       "684       45          Z-LLNle-CHO   \n",
       "688     1050             ZM447439   \n",
       "690      223              ZSTK474   \n",
       "691      266           Zibotentan   \n",
       "\n",
       "                                              Synonyms  \\\n",
       "0                          5Z-7-Oxozeaenol, LL-Z1640-2   \n",
       "4                                                 5-FU   \n",
       "20                                          KIN001-139   \n",
       "21                                          KIN001-111   \n",
       "29   AICAR, N1-(b-D-Ribofuranosyl)-5-aminoimidazole...   \n",
       "..                                                 ...   \n",
       "683                               YM-201636, YM 201636   \n",
       "684    Z-L-Norleucine-CHO, Gamma-Secretase Inhibitor 1   \n",
       "688                               ZM-447439, ZM 447439   \n",
       "690                     KIN001-167, ZSTK-474, ZSTK 474   \n",
       "691                                    ZD4054, ZD-4054   \n",
       "\n",
       "                           Targets       Target pathway   PubCHEM  Datasets  \\\n",
       "0                             TAK1       Other, kinases   9863776     GDSC1   \n",
       "4       Antimetabolite (DNA & RNA)                Other      3385     GDSC1   \n",
       "20                AKT1, AKT2, AKT3  PI3K/MTOR signaling  10172943     GDSC1   \n",
       "21                        LCK, FYN       Other, kinases   9549184     GDSC1   \n",
       "29                    AMPK agonist           Metabolism     65110     GDSC1   \n",
       "..                             ...                  ...       ...       ...   \n",
       "683                        PIKFYVE  PI3K/MTOR signaling   9956222     GDSC1   \n",
       "684                gamma-secretase                Other  16760646     GDSC1   \n",
       "688                   AURKA, AURKB              Mitosis   9914412     GDSC1   \n",
       "690                 PI3K (class 1)  PI3K/MTOR signaling  11647372     GDSC1   \n",
       "691  Endothelin-1 receptor (EDNRA)                Other   9910224     GDSC1   \n",
       "\n",
       "      number of cell lines  Screening site  \n",
       "0                      899          SANGER  \n",
       "4                      907             MGH  \n",
       "20                     397             MGH  \n",
       "21                     397             MGH  \n",
       "29                     925          SANGER  \n",
       "..                     ...             ...  \n",
       "683                    912             MGH  \n",
       "684                    397             MGH  \n",
       "688                    933          SANGER  \n",
       "690                    911             MGH  \n",
       "691                    912             MGH  \n",
       "\n",
       "[208 rows x 9 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COSMIC ID</th>\n",
       "      <th>number of drugs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>9.410000e+02</td>\n",
       "      <td>941.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>9.918890e+05</td>\n",
       "      <td>346.339001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2.213118e+05</td>\n",
       "      <td>52.803075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>6.836670e+05</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>9.068040e+05</td>\n",
       "      <td>342.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>9.097250e+05</td>\n",
       "      <td>350.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.240146e+06</td>\n",
       "      <td>386.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.674021e+06</td>\n",
       "      <td>403.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          COSMIC ID  number of drugs\n",
       "count  9.410000e+02       941.000000\n",
       "mean   9.918890e+05       346.339001\n",
       "std    2.213118e+05        52.803075\n",
       "min    6.836670e+05         1.000000\n",
       "25%    9.068040e+05       342.000000\n",
       "50%    9.097250e+05       350.000000\n",
       "75%    1.240146e+06       386.000000\n",
       "max    1.674021e+06       403.000000"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_line.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug Id</th>\n",
       "      <th>number of cell lines</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>208.000000</td>\n",
       "      <td>208.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>500.076923</td>\n",
       "      <td>806.740385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>451.290938</td>\n",
       "      <td>198.218807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>382.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>162.000000</td>\n",
       "      <td>864.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>272.500000</td>\n",
       "      <td>906.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1021.250000</td>\n",
       "      <td>914.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1530.000000</td>\n",
       "      <td>935.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Drug Id   number of cell lines\n",
       "count   208.000000             208.000000\n",
       "mean    500.076923             806.740385\n",
       "std     451.290938             198.218807\n",
       "min       1.000000             382.000000\n",
       "25%     162.000000             864.750000\n",
       "50%     272.500000             906.500000\n",
       "75%    1021.250000             914.000000\n",
       "max    1530.000000             935.000000"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs.describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pratico_sib",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
