{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intelligent systems for bioinformatics- Group 1\n",
    "\n",
    "This work is developed in the ambit of curricular unit intelligent systems for bioinformatics of the Bioinformatic Master by:\n",
    "- [Beatriz Santos](https://github.com/beatrizsantos1607)\n",
    "- [Duarte Velho](https://github.com/duartebred)\n",
    "- [Ricardo Oliveira](https://github.com/ricardofoliveira61)\n",
    "- [Rita Nobrega](https://github.com/ritanobrega00)\n",
    "- [Rodrigo Esperança](https://github.com/esperancaa)\n",
    "\n",
    "This work consists in the analysis of a dataset through the utilization of machine learning algorithms, recurring to Python as the programming language.\n",
    "The entire analysis is present on a Jupyter Notebook, organized in sections (explained later on) containing succinct explanations of the procedures and decisions taken throughout the analysis.\n",
    "\n",
    "For this work we selected the [GDSC1](https://tdcommons.ai/multi_pred_tasks/drugres) dataset. This dataset contains the wet lab IC50 for 208 drugs in 1000 cancer cells lines and can be used to design models that can predict drug response since the same compound can have differents levels of responses in different patients. With this we aim to design a model that given a pair of drug and cell line genomics profile can predict the drug response and find the best drug to treat certain patient. In this dataset the RMD normalized gene expression was used for cancer lines and the SMILES for drugs. Y is the log normalized IC50.\n",
    "\n",
    "## Notebook sections\n",
    "### 1.Preprocessing and data exploration\n",
    "- Review of all documentation available about the dataset\n",
    "- Load the dataset and realize a exploratory analysis \n",
    "- Prepare the dataset with the generation and selection of features and treatment of the missing values \n",
    "\n",
    "This stage corresponds to the 1st section of the Notebook where:\n",
    "- The dataset must the described according to the documentation\n",
    "- Summarize the characteristics of the data trought an exploratory analysis\n",
    "- Description of the preprocessing steps justifying the choices\n",
    "- Include graphics that represent the main characteristics of the dataset\n",
    "\n",
    "### 2. Non-supervised learning\n",
    "- Utilization of the adequate visualization and dimensionality reduction technique\n",
    "-  Application of clustering methods\n",
    "\n",
    "This stage corresponds to the section 2 of the Notebook where:\n",
    "- The results must be analyzed and the procedures explain\n",
    "\n",
    "### 3. Machine Learning\n",
    "- Compare the behavior of different models/methods of machine learning through the calculation of the performance metrics\n",
    "- Present the best model for the dataset\n",
    "\n",
    "This stage correspond to the section 3 of the notebook and all the results must be reported and analyzed in a critical way\n",
    "\n",
    "### 4. Deep Leaning\n",
    "- Utilization of deep learning methods similarly to the stage 3\n",
    "\n",
    "This stage correspond to the section 4 of the notebook and must report the results and have a critical analysis.\n",
    "\n",
    "All the packages used during this work are listed on the cell below\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tdc.multi_pred import DrugRes\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Preprocessing and data exploration\n",
    " The first stage of this work consists of describing the dataset that is going to be used allong the project. This stage is crucial to understand the data, its structure, its quality, and potencial analysis. Let's start by reviewing all the avaiable documentation.\n",
    "\n",
    " ### 1.1 Documentation review and dataset description\n",
    " Recent studies have shown that alterations in cancer genomes influence the clinical response to anticancer therapies. Nowadays, genomic changes are used as molecular biomarkers to identify patientes most likely to benefit from a treatment, however many cancer drugs in development or already in use have not been linked to a specific genomic markers that could guide their clinical use to diminuish the time needed to treat a patient. \n",
    "\n",
    " The discovery of cancer genome as a potencial biomarker was only possible due to the advances in the recent years in high-throughput technologies, in particular, DNA sequencing technoliges, that allow the sequencing on a scale that was previously unthinkable. To explore the increased knowledge of cancer genomics, preclinical studies that link the genomic complexity of cancer with functional readouts such as drug sensitivity are required. For this studies, cancer lines derived from many differnt naturally occuring cancer types are essensial to mimic the tissue type and genomic context of the cancer and they also provide a easy system for experimental manipulation for molecular biology and drug discovery. For this reason, several studies have used cancer cell lines to link pharmacological data with genomic information and helped define therapeutic biomarkers as well as to demonstrate that pharmacogenomic profilling in cancer cell lines can be used as a biomarker discovery platform to guide the development of new cancer therapies.\n",
    "\n",
    "The Genomics of Drug sensitivity in Cancer database, or GDSC for short, was designed to facilitate the study and understanding of the molecular features that influence drug response in cancer cell lines. The database holds datasets of drug sensitivity in cancer cells and links these data to detailed genomic information to facilitate the discovery of molecular biomarkers for drug response. This efforts are expected to, in a near future, provide a complete description of the genomic changes that occur in many cancer types\n",
    "and profund insights into the origins, evolution and progression of cancer.\n",
    "\n",
    "In order to download and load the dataset, we will use the `DrugRes` class from the `tdcommons.multi_pred` package. This package will facilitate data access to perform an initial analysis of the dataset, including its structure, format and content. The code presented below was used to download/load the dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n",
      "Loading...\n",
      "Done!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug_ID</th>\n",
       "      <th>Drug</th>\n",
       "      <th>Cell Line_ID</th>\n",
       "      <th>Cell Line</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>[3.23827250519154, 2.98225419469807, 10.235490...</td>\n",
       "      <td>2.395685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>ES3</td>\n",
       "      <td>[8.690197905033282, 3.0914731119366, 9.9924871...</td>\n",
       "      <td>3.140923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>ES5</td>\n",
       "      <td>[8.233101127037282, 2.82468731112752, 10.01588...</td>\n",
       "      <td>3.968757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>ES7</td>\n",
       "      <td>[8.33346622426757, 3.9667571228514302, 9.79399...</td>\n",
       "      <td>2.692768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>[8.39134072442845, 2.9683601858810698, 10.2606...</td>\n",
       "      <td>2.478678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177305</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-1040</td>\n",
       "      <td>[8.65368534780164, 2.9238748715081, 10.1278774...</td>\n",
       "      <td>5.353963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177306</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-407</td>\n",
       "      <td>[8.57966425274312, 2.77877087774424, 9.7680113...</td>\n",
       "      <td>4.820567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177307</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-61</td>\n",
       "      <td>[8.077115751588071, 2.78132536810578, 10.03805...</td>\n",
       "      <td>5.785978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177308</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-81</td>\n",
       "      <td>[7.7976988637889, 2.6408995410198797, 9.463400...</td>\n",
       "      <td>5.393454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177309</th>\n",
       "      <td>PFI-3</td>\n",
       "      <td>C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...</td>\n",
       "      <td>SNU-C5</td>\n",
       "      <td>[7.606903479683809, 3.15731861619302, 10.15314...</td>\n",
       "      <td>5.099131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>177310 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Drug_ID                                               Drug  \\\n",
       "0       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "1       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "2       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "3       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "4       Erlotinib  COCCOC1=C(C=C2C(=C1)C(=NC=N2)NC3=CC=CC(=C3)C#C...   \n",
       "...           ...                                                ...   \n",
       "177305      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177306      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177307      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177308      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "177309      PFI-3  C1[C@@H]2CN([C@H]1CN2C3=CC=CC=N3)/C=C/C(=O)C4=...   \n",
       "\n",
       "       Cell Line_ID                                          Cell Line  \\\n",
       "0            MC-CAR  [3.23827250519154, 2.98225419469807, 10.235490...   \n",
       "1               ES3  [8.690197905033282, 3.0914731119366, 9.9924871...   \n",
       "2               ES5  [8.233101127037282, 2.82468731112752, 10.01588...   \n",
       "3               ES7  [8.33346622426757, 3.9667571228514302, 9.79399...   \n",
       "4             EW-11  [8.39134072442845, 2.9683601858810698, 10.2606...   \n",
       "...             ...                                                ...   \n",
       "177305     SNU-1040  [8.65368534780164, 2.9238748715081, 10.1278774...   \n",
       "177306      SNU-407  [8.57966425274312, 2.77877087774424, 9.7680113...   \n",
       "177307       SNU-61  [8.077115751588071, 2.78132536810578, 10.03805...   \n",
       "177308       SNU-81  [7.7976988637889, 2.6408995410198797, 9.463400...   \n",
       "177309       SNU-C5  [7.606903479683809, 3.15731861619302, 10.15314...   \n",
       "\n",
       "               Y  \n",
       "0       2.395685  \n",
       "1       3.140923  \n",
       "2       3.968757  \n",
       "3       2.692768  \n",
       "4       2.478678  \n",
       "...          ...  \n",
       "177305  5.353963  \n",
       "177306  4.820567  \n",
       "177307  5.785978  \n",
       "177308  5.393454  \n",
       "177309  5.099131  \n",
       "\n",
       "[177310 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# download of the dataset and conversion to pandas.dataframe\n",
    "data = DrugRes(name = 'GDSC1')\n",
    "dataframe = data.get_data()\n",
    "dataframe\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The dataset donwload contains\n",
    "\n",
    "\n",
    "DATABASE CONTENT\n",
    "\n",
    "The GDSC database is based on three types of datasets as described in the following sections.\n",
    "Cell line drug sensitivity data\n",
    "\n",
    "Cancer cell line drug sensitivity data are generated from ongoing high-throughput screening performed by the Cancer Genome Project at the Wellcome Trust Sanger Institute (WTSI) and the Center for Molecular Therapeutics at Massachusetts General Hospital using a collection of >1000 cell lines (7). Compounds selected for screening are anticancer therapeutics encompassing both targeted agents and cytotoxic chemotherapeutics. They are comprised of approved drugs used in the clinic, drugs undergoing clinical development and in clinical trials and tool compounds in early phase development. They cover a wide range of targets and processes implicated in cancer biology including receptor tyrosine kinase signalling, cell cycle control, DNA damage response and the cytoskeleton. Compounds are sourced from commercial vendors or provided by collaborators in academia, biotech and the pharmaceutical industry.\n",
    "\n",
    "Cell line drug sensitivity is measured using fluorescence-based cell viability assays following 72 h of drug treatment. Dose–response curves are fitted to fluorescence signal intensities over nine drug concentrations (2-fold dilution series) to derive a multi-parameter signature of drug response. Values reported on the website include the half maximal inhibitory concentration (IC50), the slope of the dose–response curve and the area under the curve for each experiment.\n",
    "\n",
    "The current release of GDSC (release 2, July 2012) includes drug sensitivity data for 138 anticancer compounds screened across a range of 329–668 cell lines per drug (mean = 525 cell lines per drug) representing 73 169 cell line–drug interactions. This is the largest public resource available on drug sensitivity in cancer cells. Screening is ongoing and the objective is to screen these compounds, as well as additional compounds in the future, across the entire collection of >1000 cell lines. Data release occurs every 4 months and with each release, these results are updated with new data for existing drugs, as well as data for newly screened drugs.\n",
    "Genomic datasets for cell lines\n",
    "\n",
    "The total collection available for screening includes >1000 different cancer cell lines. These have been selected to represent the spectrum of common and rare types of adult and childhood cancers of epithelial, mesenchymal and haematopoietic origin. The cell lines have been extensively genomically characterized as part of the cancer cell line project from the Cancer Genome Project at the WTSI. The genomic datasets currently available for each cell line include information on somatic mutations in 75 cancer genes, genome wide gene copy number for amplification and deletion, targeted screening for seven gene rearrangements, markers of microsatellite instability, tissue type and transcriptional data. Using various statistical approaches as described below, genomic datasets are used together with drug sensitivity data for each cell line to identify genomic biomarkers of drug response. Genomic datasets within GDSC are obtained and updated directly from the Catalogue of Somatic Mutations in Cancer (COSMIC) database, a comprehensive freely available resource for the annotation and presentation of somatic mutations in cancer (8).\n",
    "Analysis of genomic features of drug sensitivity\n",
    "\n",
    "An essential component of the GDSC database is the systematic integration of large-scale genomic and drug sensitivity datasets. To identify genomic markers of drug response, we currently use two complementary analytical approaches (7). A multivariate analysis of variance (MANOVA) is used to correlate drug sensitivity (IC50 values and slope of the dose–response curve) with genomic alterations in cancer including point mutations, amplifications and deletions of common cancer genes, cancer gene rearrangements and microsatellite instability. The MANOVA identifies individual genomic features associated with drug sensitivity and for each drug–gene association reports a size effect and statistical significance of the association.\n",
    "\n",
    "We also apply elastic net regression, a penalized linear modelling technique, to identify multiple interacting genomic features influencing each drug response. Genomic data used in the elastic net analysis include all of those used in the MANOVA and also incorporate genome-wide transcriptional profiles and tissue type. The elastic net selects which of these features are associated with drug response as measured by IC50 values across the cell line panel. For each drug, a feature list is built comprised of mutations, transcripts and tissue with an effect size assigned to each.\n",
    "\n",
    "A more detailed description of the different statistical analyses performed, as well as guidance on interpreting the results, can be found on the ‘Help & Documentation’ webpages under the ‘statistical analysis’ tab."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n",
      "Loading...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "17737"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_symbols = data.get_gene_symbols()\n",
    "len(gene_symbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Dataset Statistics ---\n",
      "208 unique drugs.\n",
      "958 unique cell lines.\n",
      "177310 drug-cell line pairs.\n",
      "--------------------------\n"
     ]
    }
   ],
   "source": [
    "data.print_stats()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pratico_sib",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
